lib/Sema/TypeCheckPattern.cpp

```diff
@@ -775,10 +775,31 @@ static bool validateParameterType(ParamDecl *decl, DeclContext *DC,
   return hadError;
 }
 
+/// Request nominal layout for any types that could be sources of typemetadata
+/// or conformances.
+void TypeChecker::requestRequiredNominalTypeLayoutForParameters(
+    ParameterList *PL) {
+  for (auto param : *PL) {
+    auto type = param->hasType() ? param->getType()->getCanonicalType()
+                                 : param->getTypeLoc().getType();
+    if (!type)
+      continue;
+    if (auto *generic = dyn_cast<BoundGenericType>(type.getPointer())) {
```

Comment:
... and then we'd use `type->getAs<BoundGenericType>()` here to look through sugar. You can also use `type->getAnyGeneric()` to get the generic decl automatically.